packages/@react-spectrum/s2/src/Tabs.tsx

```diff
@@ -10,9 +10,302 @@
+import {size, style} from '../style/spectrum-theme' with {type: 'macro'};
+import {useDOMRef} from '@react-spectrum/utils';
+import {useLayoutEffect} from '@react-aria/utils';
+import {useLocale} from '@react-aria/i18n';
 
+interface TabsProps extends Omit<AriaTabsProps, 'className' | 'style' | 'children'>, UnsafeStyles {
+  /** Spectrum-defined styles, returned by the `style()` macro. */
+  styles?: StylesPropWithHeight,
+  /** The content to display in the tabs. */
+  children?: ReactNode,
+  density?: 'compact' | 'regular'
+}
+
+interface TabProps extends Omit<AriaTabProps, 'children' | 'style' | 'className'> {
+  children?: ReactNode
+}
+
+interface TabListProps<T> extends Omit<AriaTabListProps<T>, 'children' | 'style' | 'className'>{
+  children?: ReactNode
+}
+
+interface TabPanelProps extends Omit<AriaTabPanelProps, 'children' | 'style' | 'className'>{
+  children?: ReactNode
+}
+
+export function TabPanel(props: TabPanelProps) {
+  return (
+    <AriaTabPanel {...props} className={style({marginTop: 4, color: 'neutral'})} />
+  );
+}
+
+const tab = style({
+  ...focusRing(),
+  display: 'inline-flex',
+  color: {
+    default: 'neutral-subdued',
+    isSelected: 'neutral',
+    isHovered: 'neutral-subdued',
+    isDisabled: 'disabled',
+    forcedColors: {
+      isSelected: 'Highlight',
+      isDisabled: 'GrayText'
+    }
+  },
+  borderRadius: 'sm',
+  gap: size(6),
+  height: {
+    density: {
+      compact: 32,
+      regular: 48
+    }
+  },
+  alignItems: 'center',
+  position: 'relative'
+});
+
+const icon = style({
+  paddingEnd: 4,
+  '--iconPrimary': {
+    type: 'fill',
+    value: 'currentColor'
+  }
+});
+
+export function Tab(props: TabProps) {
+  let {density} = useContext(TabsInternalContext);
+
+  return (
+    <RACTab 
+      {...props}
+      className={renderProps => tab({...renderProps, density})}>
+      <Provider
+        values={[
+          [IconContext, {styles: icon}]
+        ]}>
+        {props.children}
+      </Provider>
+    </RACTab>
+  );
+}
+
+const tablist = style({
+  display: 'flex',
+  height: {
+    orientation: {
+      horizontal: {
+        density: {
+          compact: 32,
+          regular: 48
+        }
+      }
+    }
+  },
+  gap: {
+    orientation: {
+      horizontal: {
+        density: {
+          compact: 24,
+          regular: 32
+        }
+      }
+    }
+  },
+  flexDirection: {
+    orientation: {
+      vertical: 'column'
+    }
+  },
+  paddingEnd: {
+    orientation: {
+      vertical: 20
+    }
+  },
+  paddingStart: {
+    orientation: {
+      vertical: 12
+    }
+  },
+});
+
+export function TabList<T extends object>(props: TabListProps<T>) {
+  let {density, isDisabled, disabledKeys, orientation = 'horizontal'} = useContext(TabsInternalContext);
+  let state = useContext(TabListStateContext);
+  let [selectedTab, setSelectedTab] = useState<HTMLElement | undefined>(undefined);
+  let tablistRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {    
```

Comment:
I think this should be added to `useTagGroup, I see previously we handled it in RSP Tabs.
Looks like it's dependent on rendering, however, we could probably make use of the collection/selectionManager instead of querying the dom. I have no idea why we were doing that.

I'm ok for it to be followup as well